abn/service.go

```diff
@@ -0,0 +1,150 @@
+	WATCHER_CONFIG_ENV = "WATCHER_CONFIG"
+)
+
+var (
+	// port the service listens on
+	port = flag.Int("port", 50051, "The server port")
+)
+
+// Start is entry point to configure services and start them
+func Start(kd *driver.KubeDriver) {
+	// Initialize kubernetes driver
+	if err := kd.Init(); err != nil {
+		log.Logger.Fatal("unable to initialize kubedriver")
+	}
+	// Initialize appliction map with ReaderWriter
+	abnapp.Applications.SetReaderWriter(&abnapp.ApplicationReaderWriter{Client: kd.Clientset})
+
+	// read abn config (resources and namespaces to watch)
+	abnConfigFile, ok := os.LookupEnv(WATCHER_CONFIG_ENV)
+	if !ok {
+		log.Logger.Fatal("configuation file is required")
+	}
+
+	stopCh := make(chan struct{})
+
+	// set up resource watching as defined by config
+	abnConfig := watcher.ReadConfig(abnConfigFile)
+	w := watcher.NewIter8Watcher(kd, abnConfig.Resources, abnConfig.Namespaces)
+	go w.Start(stopCh)
+
+	// launch gRPC server to respond to frontend requests
+	go launchGRPCServer([]grpc.ServerOption{}, kd)
+
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGTERM, os.Interrupt)
+
+	<-sigCh
+	close(stopCh)
+}
+
+// newServer returns a new gRPC server
+func newServer(kd *driver.KubeDriver) *abnServer {
+	s := &abnServer{
+		Driver: kd,
+	}
+	return s
+}
+
+type abnServer struct {
+	Driver *driver.KubeDriver
+	pb.UnimplementedABNServer
+}
+
+// Lookup identifies a track that should be used for a given user
+// This method is exposed to gRPC clients
+func (server *abnServer) Lookup(ctx context.Context, appMsg *pb.Application) (*pb.Session, error) {
+	abnapp.Applications.Lock()
+	defer abnapp.Applications.Unlock()
+
+	track, err := pb.Lookup(appMsg.GetName(), appMsg.GetUser())
+	if err != nil || track == nil {
+		return nil, err
+	}
+
+	return &pb.Session{
+		Track: *track,
+	}, err
+}
+
+func (server *abnServer) WriteMetric(ctx context.Context, metricMsg *pb.MetricValue) (*emptypb.Empty, error) {
+	abnapp.Applications.Lock()
+	defer abnapp.Applications.Unlock()
+
+	a, err := abnapp.Applications.Get(metricMsg.Application, false)
+	if err != nil || a == nil {
+		return &emptypb.Empty{}, errors.New("unexpected: cannot find record of application " + metricMsg.GetApplication())
+	}
+
+	track, err := pb.Lookup(metricMsg.GetApplication(), metricMsg.GetUser())
+	if err != nil || track == nil {
+		return &emptypb.Empty{}, err
+	}
+
+	version, ok := a.Tracks[*track]
+	if !ok {
+		return &emptypb.Empty{}, errors.New("track not mapped to version")
+	}
+
+	v, _ := a.GetVersion(version, false)
+	if v == nil {
+		return &emptypb.Empty{}, errors.New("unexpected: trying to write metrics for unknown version")
+	}
+
+	value, err := strconv.ParseFloat(metricMsg.GetValue(), 64)
+	if err != nil {
+		log.Logger.Warn("Unable to parse metric value ", metricMsg.GetValue())
+		return &emptypb.Empty{}, nil
+	}
+
+	m, _ := v.GetMetric(metricMsg.GetName(), true)
+	m.Add(value)
+
+	// persist updated metric
+	abnapp.Applications.Write(a)
```

Comment:
Update this secret only if the last update to this secret was 60 seconds ago.

Reply:
Create a BatchedWrite(?) method for use in WriteMetrics. 
***
When the application object gets too big, we drop untracked versions.
***
There is a limit on number of metrics-per-version.

Reply:
Probably should make the timeout value be user configurable at install time.

Reply:
How do I compute the "size" of a secret?

Reply:
set maximum length of byte array of content of application data to 750,000